### PR TITLE
Distributed Sampling in cuGraph-PyG

### LINKED ISSUE #3383: [IMP] Don't Copy to Host When Creating CuGraphStore (cugraph-pyg)
Currently, regardless of input type of the edge indices, cugraph-pyg will convert to a numpy array, and then later on, copy the numpy array back to the device.  This can be avoided.

### LINKED ISSUE #3836: Remove Deprecated Homogeneous Renumbering From CuGraphStore in cuGraph-PyG
Renumbering for homogeneous graphs has been moved to the sampling side, so the logic to renumber homogeneous graphs should be removed.

### LINKED ISSUE #3942: [BUG] CSR/CSC Output from Bulk Sampler is unusable in cuGraph-PyG
Currently, the CSR/CSC output from the bulk sampler is unusable since the number of previous destinations does not match the number of sources.  This is most likely because something is going wrong with out-degree 0 vertices.  Further investigation is needed to identify the source of the problem.

### LINKED ISSUE #4051: [IMP] cuGraph-PyG Should Use `torch_geometric.EdgeIndex`
`torch_geometric.EdgeIndex` natively supports caching CSR/CSC formats and is a better maintained alternative to `torch_sparse`.  We should migrate to this new class for the outputs of cuGraph-PyG sampling.

https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.edge_index.EdgeIndex.html#torch_geometric.edge_index.EdgeIndex

### LINKED ISSUE #4202: [FEA] Use PyG Interfaces for Modifying Feature Store
Unlike cuGraph-DGL, cuGraph-PyG currently lacks support for using the native PyG interface for feature management.  Instead, users have to construct and pass a cuGraph `FeatureStore` object, which exposes implementation details and reduces usability.

We should implement the `FeatureStore` interface properly as well as add functions for constructing a `CuGraphStore` from a PyG object.

### LINKED ISSUE #4248: [FEA] New Distributed Graph Store for cuGraph-PyG
This new distributed graph store will not use Dask. Instead, each worker will pass in a slice of the edgelist, and a pylibcugraph.MGGraph will be created on each worker. This store should also be completely unaware of feature storage. It will be fully compatible with the distributed sampler (https://github.com/rapidsai/cugraph/issues/4246).

### LINKED ISSUE #4249: [FEA] Split cuGraph-PyG Feature Store from Graph Store
`cuGraph-PyG` should follow the universal paradigm of feature stores being separate from the graph store.  This is critical for enabling a variety of user workflows which are currently broken with the current interface.  It is also critical in simplifying the integration of `WholeGraph` into `cuGraph-PyG`.

### LINKED ISSUE #4252: New Distributed MG cuGraph GNN Examples
Need to create new examples that use the new distributed sampling workflow.

### LINKED ISSUE #4326: [FEA] Support Isolated Vertices in cuGraph-PyG/DGL
We currently are unable to handle isolated vertices for GNN sampling, since those are dropped during graph creation.  This is causing issues for several users, who are unable to train their graphs with many isolated vertices.  Adding self loops is not an option since those will cause a major accuracy drop for heterogeneous graphs and may also violate the schema of the users' upstream graph storage.

### PR DESCRIPTION
**Pull request by alexbarghi-nv**
Distributed sampling in cuGraph-PyG.  Also renames the existing API to clarify that it is dask based.
Adds a dependency on `tensordict` for `cuGraph-PyG` which supports the new `TensorDictFeatureStore`.
Also no longer installs `torch-cluster` and `torch-spline-conv` in CI for testing since that results in an `ImportError` and neither of those packages are needed.

Requires PyG 2.5.  Should be merged after #4335 

Merge after #4355 

Closes #4248 
Closes #4249 
Closes #3383 
Closes #3942 
Closes #3836 
Closes #4202 
Closes #4051 
Closes #4326 
Closes #4252 
Partially addresses #3805

### CODE REVIEW

File: python/cugraph-pyg/cugraph_pyg/examples/README.md
Comment (alexbarghi-nv) [inline python/cugraph-pyg/cugraph_pyg/examples/README.md:1]: The content of this file was migrated to `graph_sage_mg.py`

File: python/cugraph-pyg/cugraph_pyg/examples/cugraph_dist_sampling_sg.py
Comment (alexbarghi-nv) [inline python/cugraph-pyg/cugraph_pyg/examples/cugraph_dist_sampling_sg.py:1]: CSR is the fastest compression based on benchmarking

File: python/cugraph-pyg/cugraph_pyg/examples/gcn_dist_snmg.py
Comment (alexbarghi-nv) [inline python/cugraph-pyg/cugraph_pyg/examples/gcn_dist_snmg.py:1]: The MNMG example will be added along with the new WG feature store, and will use the new WG feature store to avoid replicating data across processes.

File: python/cugraph-pyg/cugraph_pyg/data/graph_store.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+                        self.__edge_indices[edge_attr.edge_type][1].max() + 1
+                    )
+
+        if self.is_multi_gpu:
+            vtypes = num_vertices.keys()
+            for vtype in vtypes:
+                sz = torch.tensor(num_vertices[vtype], device="cuda")
+                torch.distributed.all_reduce(sz, op=torch.distributed.ReduceOp.MAX)
+                num_vertices[vtype] = int(sz)
+        return num_vertices
```
Comment (tingyu66) [inline python/cugraph-pyg/cugraph_pyg/data/graph_store.py:201]: Should we cache `num_vertices` and `num_edges` since they require collective communications?

File: python/cugraph-pyg/cugraph_pyg/data/graph_store.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+            raise ValueError("Only COO format supported")
+
+        if isinstance(edge_index, (cupy.ndarray, cudf.Series)):
+            edge_index = torch.as_tensor(edge_index, device="cuda")
+        elif isinstance(edge_index, (np.ndarray)):
+            edge_index = torch.as_tensor(edge_index, device="cpu")
+        elif isinstance(edge_index, pandas.Series):
+            edge_index = torch.as_tensor(edge_index.values, device="cpu")
+        elif isinstance(edge_index, cudf.Series):
+            edge_index = torch.as_tensor(edge_index.values, device="cuda")
```
Comment (tingyu66) [inline python/cugraph-pyg/cugraph_pyg/data/graph_store.py:72]: What's the reason of creating CPU tensor from numpy and pandas input? Is it because we want to delay data transfer until graph creation? If so, I think we should use `device=edge_index.device` in other cases.

**Review by tingyu66 — APPROVED** [review]
Approved with some comments and questions. The overall logic looks reasonable and trust you with matching the detailed interface of PyG's native Graph- and FeatureStore.

**Review by BradReesWork — APPROVED** [review]

File: python/cugraph-pyg/cugraph_pyg/data/graph_store.py (hunk 2 above)
Comment (alexbarghi-nv) [reply python/cugraph-pyg/cugraph_pyg/data/graph_store.py:72]: I don't think there is an easy way to get the device of a cudf dataframe here

File: python/cugraph-pyg/cugraph_pyg/examples/gcn_dist_snmg.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+    kwargs = dict(
+        num_neighbors=[fan_out] * num_layers,
+        batch_size=batch_size,
+    )
+    # Set Up Neighbor Loading
+    from cugraph_pyg.data import GraphStore, TensorDictFeatureStore
+    from cugraph_pyg.loader import NeighborLoader
+
+    graph_store = GraphStore(is_multi_gpu=True)
+    ixr = torch.tensor_split(data.edge_index, world_size, axis=1)[rank]
```
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply python/cugraph-pyg/cugraph_pyg/examples/gcn_dist_snmg.py]: fixed

File: python/cugraph-pyg/cugraph_pyg/examples/gcn_dist_snmg.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+                args.fan_out,
+                split_idx,
+                dataset.num_classes,
+                wall_clock_start,
+                tempdir,
+                args.num_layers,
+            ),
+            nprocs=world_size,
+            join=True,
+        )
```
[the remark this answers is not in the document]
Comment (alexbarghi-nv) [reply python/cugraph-pyg/cugraph_pyg/examples/gcn_dist_snmg.py:322]: fixed

File: python/cugraph-pyg/cugraph_pyg/data/graph_store.py (hunk 2 above)
Comment (tingyu66) [reply python/cugraph-pyg/cugraph_pyg/data/graph_store.py:72]: I see. It makes sense.

File: python/cugraph-pyg/cugraph_pyg/data/graph_store.py (hunk 1 above)
Comment (alexbarghi-nv) [reply python/cugraph-pyg/cugraph_pyg/data/graph_store.py:201]: We cache the offsets, which are more frequently used, but I might add caching the vertex and edge counts if they are used more often.  So maybe in the future.

**Review by AyodeAwe — APPROVED** [review]
